### ISSUE TITLE
New l3draw functions to extract x/y/z-coordinate of a point

### ISSUE THREAD

**Issue by muzimuzhi** [latex3/latex3#1491]
Discussed in https://github.com/latex3/latex3/issues/1488#issuecomment-1974746683.

`pgf` provides `\pgfextractx` and `\pgfextracty` with syntax `\pgfextractx{<dimension>}{<point>}`, see https://tikz.dev/base-points#sec-101.6.

One possibility: `\draw_point_extract_x:Nn <dim var> { <point> }`.

If needed `\draw_point_extract_last_x:N` which extracts the x-coordinate of last updated point can also be provided. `pgf` provides `\pgfgetlastxy{<macro for x>}{<macro for y>}`.

**Comment by josephwright** [comment]
See comments in the sources: the are deliberately not includes pending solid use cases. I suspect that 'local' auxiliaries are likey the base plan.

**Comment by muzimuzhi** [comment]
https://github.com/latex3/latex3/blob/2ea12efc6a163bf1ae3fd0ad6d05dabc3701cf2a/l3experimental/l3draw/l3draw-points.dtx#L67-L78

**Comment by muzimuzhi** [comment]
> ```
> %   \item \cs{pgfextractx}, \cs{pgfextracty}: Available by applying
> %     \cs{use_i:nn}/\cs{use_ii:nn} or similar to the \texttt{e}-type
> %     expansion of a point expression.
> ```

But public `\draw_point_...` functions expand points to points themselves, aka two-element comma-separated lists. Only the internal `\__draw_point_process:nn` expands a point to `{<x-coord>}{<y-coord>}`, which then can be passed to `\use_i:nn`/`\use_ii:nn` to extract x/y-coordinate.

That's why @jasperhabicht uses `\clist_item:en` in the demonstration in https://github.com/latex3/latex3/issues/1488#issuecomment-1974746683.

**Comment by josephwright** [comment]
@muzimuzhi  Yes, my code comment is inaccurate but the general idea stands. In `pgf` you need the 'extract' commands as the information is internal. In `l3draw` we can do everything using an fp expression by expansion, so can easily create an 'extract the value' command. But more widely, there are two things. First, as we can use tuples throughout, I'm not sure point extraction is very useful - we don't need to access just-x or just-y. Second, as this is really a general tuple feature, I suspect it should be handled within `l3fp`, something like `\fp_eval:n { select(2, (<tuple>) }` -> y-value.

**Comment by josephwright** [comment]
(Whole reason this code is in `l3experimental` is of course that this sort of stuff needs exploring.)

**Comment by josephwright** [comment]
BTW, points don't have a z-co-ordinate ;)

**Comment by car222222** [comment]
But this is l3draw, so surely it is in 3D?? :-)

**Comment by jasperhabicht** [comment]
I think there can be situations where you need just the *x* or the *y* part of a point coordinate, for example if you want to calculate trigonometric stuff where you can't use tuples (for example when using `atand()`, you need both values). 

Still, I agree that extracting a value from a tuple should rather be a feature of `l3fp`.

**Comment by jasperhabicht** [comment]
Actually, mbert came up with a great idea [in the TeX.SX chat](https://chat.stackexchange.com/transcript/message/66844329#66844329) how to easily do this using a custom function which works nicely nere because we have always 2-tuples for coordinates. We can easily define

```tex
\fp_new_function:n { drawgetx }
\fp_set_function:nnn { drawgetx } { x , y } { x }

\fp_new_function:n { drawgety }
\fp_set_function:nnn { drawgety } { x , y } { y }
```

and then use the result in some `\fp_eval:n` or similar context as `drawgetx( <some coordinate> )`.

**Comment by jasperhabicht** [comment]
For the record: I currently use the following approach in `hawkdraw` which proves quite robust:

```tex
\cs_new:Npn \hawkdraw_tuple_use_i:n #1 {
    \fp_to_dim:n {
        \exp_last_unbraced:Ne
        \__hawkdraw_tuple_use_i_aux:w { \fp_eval:n {#1} }
    }
}

\cs_new:Npn \__hawkdraw_tuple_use_i_aux:w ( #1 , #2 ) {#1}

\cs_new:Npn \hawkdraw_tuple_use_ii:n #1 {
    \fp_to_dim:n {
        \exp_last_unbraced:Ne
        \__hawkdraw_tuple_use_ii_aux:w { \fp_eval:n {#1} }
    }
}

\cs_new:Npn \__hawkdraw_tuple_use_i_aux:w ( #1 , #2 ) {#2}
```
